src/Microsoft.AspNet.Mvc.Xml/RequiredValidationHelper.cs

```diff
@@ -0,0 +1,182 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+using System.Runtime.Serialization;
+using Microsoft.AspNet.Mvc.ModelBinding;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    /// <summary>
+    /// Validates types having value type properties decorated with <see cref="RequiredAttribute"/>
+    /// but no <see cref="DataMemberAttribute"/>.
+    /// </summary>
+    public class DataAnnotationRequiredAttributeValidation
+    {
+        private ConcurrentDictionary<Type, Dictionary<Type, List<string>>> cachedValidationErrors
+            = new ConcurrentDictionary<Type, Dictionary<Type, List<string>>>();
+
+        public void Validate([NotNull] Type modelType, [NotNull] ModelStateDictionary modelStateDictionary)
+        {
+            var visitedTypes = new HashSet<Type>();
+
+            // Every node maintains a dictionary of Type => Errors. 
+            // It's a dictionary as we want to avoid adding duplicate error messages.
+            // Example:
+            // Type 'Store' has properties of type 'Address' and list of 'Employee'.
+            // 'Employee' type also has a property of type 'Address'.
+            // In this case the validation errors from the perspective of the type 'Store'
+            // should not have duplicate errors for 'Address'.
```

Comment:
Don't be afraid to just type in code when making an example. It takes up more space but it reads much more clearly.